cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java

```diff
@@ -0,0 +1,228 @@
+   * control metadata will be null when the application is deployed in the namespace. It will be
+   * non-null when the application is pulled from the remote repository and deployed in the
+   * namespace.
+   *
+   * @param appId             {@link ApplicationId} for which the source control metadata is being
+   *                          set.
+   * @param sourceControlMeta The {@link SourceControlMeta} to be set. Can be {@code null} if
+   *                          application is just deployed.
+   * @throws IOException If failed to write the data.
+   */
+  public void setNamespaceSourceControlMeta(ApplicationId appId,
+      @Nullable SourceControlMeta sourceControlMeta)
+      throws IOException {
+    StructuredTable scmTable = getNamespaceSourceControlMetadataTable();
+    scmTable.upsert(getNamespaceSourceControlMetaFields(appId, sourceControlMeta));
+  }
+
+  /**
+   * Updates the source control metadata of an application in a namespace. This function is called
+   * when application is pushed from the namespace to a remote repository. So,
+   * {@link SourceControlMeta} for this application should not be null and is updated in the
+   * {@code NamespaceSourceControlMetadata} table
+   *
+   * @param appId   {@link ApplicationId} whose source control metadata is being updated.
+   * @param scmMeta The source control metadata to update.
+   * @throws IOException                  if failed to write the metadata
+   * @throws ApplicationNotFoundException If the specified application ID is not found.
+   */
+  public void updateNamespaceSourceControlMeta(ApplicationId appId, SourceControlMeta scmMeta)
+      throws IOException, ApplicationNotFoundException {
+    StructuredTable namespaceSourceControlMetadataTable = getNamespaceSourceControlMetadataTable();
+    List<Field<?>> fields = getPrimaryKey(appId);
+    Optional<StructuredRow> existing = namespaceSourceControlMetadataTable.read(fields);
+    if (!existing.isPresent()) {
+      throw new ApplicationNotFoundException(appId);
+    }
+    fields.add(Fields.stringField(
+        StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD,
+        scmMeta.getFileHash()));
+    fields.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD,
+            scmMeta.getCommitId()));
+    fields.add(
+        Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+            scmMeta.getLastSyncedAt().toEpochMilli()));
+    namespaceSourceControlMetadataTable.upsert(fields);
+  }
+
+  /**
+   * Deletes the source control metadata associated with the specified application ID from the
+   * namespace.
+   *
+   * @param appId {@link ApplicationId} whose source control metadata is to be deleted.
+   * @throws IOException if it failed to read or delete the metadata
+   */
+  public void deleteNamespaceSourceControlMetadata(ApplicationId appId) throws IOException {
```

Comment:
don't need 'Namespace' in the method name

Reply:
Fixed